File: javascript/programmers/javascript/level3/이중우선순위큐.js
const operations = ["I 7", "I 5", "I -5", "D -1"];

const solution = (operations) => {
  let answer = [];
  for (let i = 0; i < operations.length; i++) {
    const num = Number(operations[i].substring(2));
    switch (operations[i].substring(0, 1)) {
      case "I":
        answer.push(num);
        answer.sort((a, b) => {
          return a - b;
        });
        break;
      case "D":
        if (num === 1) {
          answer.pop();
        } else {
          answer.shift();
        }
        break;
    }
  }
  if (answer.length === 0) {
    return [0, 0];
  }

  return [answer[answer.length - 1], answer[0]];
};

console.log(solution(operations));

File: python/programmers/level3/이중우선순위큐.py
from heapq import heappush, heappop


def solution(operations):
    maxHeap = []
    minHeap = []

    for operation in operations:
        if operation == 'D 1':
            if maxHeap:
                heappop(maxHeap)
                if maxHeap == [] or -maxHeap[0] < minHeap[0]:
                    minHeap = []
                    maxHeap = []
        elif operation == 'D -1':
            if minHeap:
                heappop(minHeap)
                if minHeap == [] or -maxHeap[0] < minHeap[0]:
                    maxHeap = []
                    minHeap = []
        else:
            num = int(operation[2:])
            heappush(maxHeap, -num)
            heappush(minHeap, num)

    if minHeap == []:
        return [0, 0]

    return [-heappop(maxHeap), heappop(minHeap)]
